services/api/lib/postgre.js

```diff
@@ -295,6 +295,295 @@ module.exports = function (pg) {
+                  data.type,
+                  JSON.stringify(data.attributes || {}),
+                  JSON.stringify(data.styles || {})
+                ];
+                break;
+              }
+              case 'update': {
+                values = [
+                  JSON.stringify(data.styles || {}),
+                  JSON.stringify(data.attributes || {}),
+                  data.id
+                ];
+                break;
+              }
+              case 'remove': {
+                values = [
+                  data.id
+                ];
+                break;
+              }
+            }
+          }
+        }
+        return values;
+      }
+
+      function hasPermissionsForAction(userId, type, method, data, results, transaction) {
+        return new Promise(function(resolve, reject) {
+          var lookupType;
+          var lookupId;
+          if ( method === 'create' ) {
+            switch (type) {
+              case 'projects': {
+                return resolve(true);
+              }
+              case 'pages': {
+                lookupType = 'projects';
+                lookupId = data.projectId;
+                break;
+              }
+              case 'elements': {
+                lookupType = 'pages';
+                lookupId = data.pageId;
+                break;
+              }
+            }
+          } else {
+            lookupType = type;
+            lookupId = data.id;
+          }
+
+          // coerce these values to string, because bigints come in from PG as strings
+          lookupId = '' + lookupId;
+          userId = '' + userId;
+
+          var txResult;
+          for (var i = 0; i < results.length; ++i) {
+            var result = results[i];
+            if (
+              result.method === 'create' &&
+              result.type === lookupType &&
+              result.id === lookupId )
+            {
+              txResult = result;
+              break;
+            }
+          }
+
+          // this action refers to an object created earlier in the transaction
+          if ( txResult ) {
+            return resolve(txResult.user_id === userId);
+          }
+          executeTransaction(transaction, queries[lookupType].findOneById, [lookupId])
+          .then(function(results) {
+            if ( !results.rows.length || results.rows[0].user_id !== userId ) {
+              return resolve(false);
+            }
+            resolve(true);
+          })
+          .catch(function(err) {
+            reject(err);
+          });
+        });
+      }
+
+      server.method('projects.bulk', function(actions, userId, done) {
+        var transaction;
+        var transactionResults;
+        getTransactionClient()
+        .then(function(t) {
+          transaction = t;
+          return begin(transaction);
+        })
+        .then(function() {
+          return BPromise.resolve(actions);
+        })
+        .reduce(function(results, action, actionIndex) {
```

Comment:
this is super-nesty, it feels like this could be something like
```
.then(function() {
  return BPromise.resolve(actions);
})
.reduce(authReduce);
```
with the reduce function defined as its own thing
```